sklearn/metrics/pairwise.py

```diff
@@ -933,21 +948,22 @@ def manhattan_distances(X, Y=None, *, sum_over_features=True):
     """
     X, Y = check_pairwise_arrays(X, Y)
 
+    if issparse(X) or issparse(Y):
+        X = csr_matrix(X, copy=False)
+        Y = csr_matrix(Y, copy=False)
+        X.sum_duplicates()  # this also sorts indices in-place
+        Y.sum_duplicates()
+
+    if sum_over_features and PairwiseDistances.is_usable_for(X, Y, metric="manhattan"):
+        return PairwiseDistances.compute(X, Y, metric="manhattan")
+
     if issparse(X) or issparse(Y):
         if not sum_over_features:
             raise TypeError(
                 "sum_over_features=%r not supported for sparse matrices"
                 % sum_over_features
             )
 
-        X = csr_matrix(X, copy=False)
-        Y = csr_matrix(Y, copy=False)
-        X.sum_duplicates()  # this also sorts indices in-place
-        Y.sum_duplicates()
-        D = np.zeros((X.shape[0], Y.shape[0]))
-        _sparse_manhattan(X.data, X.indices, X.indptr, Y.data, Y.indices, Y.indptr, D)
-        return D
-
     if sum_over_features:
         return distance.cdist(X, Y, "cityblock")
```

Comment:
If X and Y are lists of lists of numbers (which is do support yet something do not convert to numpy arrays), then this code will be called.

More generally, sometimes lists of lists of numbers are passed down to Cython implementations even after inputs being validated. We have not yet decided how to treat those cases.

Reply:
Thanks for the clarification.

Can't we just make sure to convert to 2D numpy arrays in this the input validation step of the public methods / functions instead?

If not or it's too far reaching to be considered being done as part of this PR maybe it would be worth adding a TODO inline comment to state that this branch is still needed for this specific case and we might want to turn those list-of-lists to numpy array arrays in the future.

Reply:
I think it's worth changing the logic regarding list-of-lists in another PR because this change would impact a lot of implementations.

In the meantime, I will just add a inline TODO comment as you suggested.